### PR TITLE
wasmd 30 + ibc v4 + IBCFees

### PR DESCRIPTION
**Pull request by Reecepbcups**
Pending ibc-go ICS23 v0.9.0 bump 2868 PR

wasmd 0.30.0
ibc go v4.2.0

New IBCFeeKeeper by following official guide
- https://ibc.cosmos.network/main/middleware/ics29-fee/integration.html

ibcRouter updated wasmStack with ibc fee keeper middleware. Same for transfer & icaHost

Added ibc v3->4 upgrade comment from official guide
- https://github.com/cosmos/ibc-go/blob/v5.1.0/docs/migrations/v3-to-v4.md#migration-to-fix-support-for-base-denoms-with-slashes

### CODE REVIEW

**Comment by faddat** [comment]
dude I love this!

I'm going to approve from the notional side.

We'll need approvals from others, though.

**Review by faddat — CHANGES_REQUESTED** [review]
I think this requires one or more upgrade handlers:

https://github.com/cosmos/ibc-go/blob/v5.1.0/docs/migrations/v3-to-v4.md

**Comment by vuong177** [comment]
@Reecepbcups should we hold it until v13? IBCFee is a new feature on IBC and I'm not sure we need it in v12.

**Comment by Reecepbcups** [comment]
@vuong177 IBC v4 & ibcfees is required for wasmd 0.30.0 - so if we want latest wasmd has to be in v12

**Comment by Reecepbcups** [comment]
@faddat I looked over this document and seems you only need the upgrade handler if your chain has factory/address/name denoms. 
Which we don't have until v12 launch with the tokenfactory

Maybe I'm wrong will look into again tomorrow

**Comment by faddat** [comment]
So I suspect that those are outdated documents. But I don't know. I'm going to reach out to the IBC team to make sure.

**Comment by faddat** [comment]
I would also like to wait to merge this until these changes land upstream:

https://github.com/cosmos/ibc-go/pull/2868

**Comment by Reecepbcups** [comment]
@faddat oops yea I had forgot to do this. It is now, done in above commit

**Comment by faddat** [comment]
Ok, so per the dev chat, we're going to wait on an alternative impl that doesn't include the ibcfee module. 

Personally, prefer this route.

**Review by faddat — DISMISSED** [review]
let's do it, but wait on upgrading uni till ibc-go changes for ics23 v0.9.0 are in

**Comment by vuong177** [comment]
`failed to load latest version: failed to load store: initial version set to 30, but found earlier version 1`

I got that err when running a testnet.

**Comment by vuong177** [comment]
```
failed to load latest version: failed to load store: initial version set to 30, but found earlier version 1
```
I got that err when running a testnet. @faddat @Reecepbcups.

**Comment by vuong177** [comment]
Hmm I got it when run the upgrade test in e2e.

**Review by faddat — CHANGES_REQUESTED** [review]
requesting changes due to reported issue

**Comment by Reecepbcups** [comment]
@vuong177 Fixed, I had IBCFees in storeUpgrade. Since its middleware it broken the app given no params

Software upgrade to v12 works now here

**Review by faddat — APPROVED** [review]

**Comment by faddat** [comment]
This PR should be prioritized for merge, so that we can properly run various tests against it in subsequent PRs.

**Comment by faddat** [comment]
To me @vuong177 it looked to me like it could be the test itself as an issue.  I just kicked off CI again to check.

**Review by JakeHartnell — APPROVED** [review]
LGTM. utACK.

**Comment by the-frey** [comment]
Looks like this needs a resolve against base branch but agree we should get this one in if poss.

**Comment by giansalex** [comment]
Can we activate StargateQueries as part of wasmd 0.30? https://github.com/CosmWasm/wasmd/pull/1069

**Comment by Reecepbcups** [comment]
> Can we enable StargateQueries as part of wasmd 0.30? https://github.com/CosmWasm/wasmd/pull/1069

@giansalex Yes, I will commit this in the morning

**Comment by the-frey** [comment]
Which queries do you want to see enabled @giansalex ?

**Comment by giansalex** [comment]
> Which queries do you want to see enabled @giansalex ?

I am looking for the routes needed to implement ICQ in a contract.

- `/ibc.core.connection.v1.Query/Connection`
- `/ibc.core.client.v1.Query/ClientState`
- `/ibc.core.client.v1.Query/ConsensusState`

File: app/keepers/keepers.go
Comment (JakeHartnell) [inline app/keepers/keepers.go:310]: Just checking and flagging, no risk of non-determinism from these correct?

Comment (JakeHartnell) [reply app/keepers/keepers.go:310]: Great.

**Comment by the-frey** [comment]
I think this is ready to ship - QQ should we have stargate queries in another PR? That feels like the sketchiest thing that we might have to back out of after testing & might be good to have it atomically within a single PR

File: app/upgrades/v12/upgrades.go
Comment (giansalex) [inline app/upgrades/v12/upgrades.go:20]: missing to add  `ibcfeetypes.ModuleName` ?
https://github.com/CosmosContracts/juno/blob/f9ae0c654e2c034cc1c71701f191421dd94b8b88/app/upgrades/v12/constants.go#L17

Comment (Reecepbcups) [reply app/upgrades/v12/upgrades.go:20]: Ahh yes thank you. I was having consensus version issues before but forgot to write the patch back in

**Review by faddat — APPROVED** [review]

**Review by faddat — APPROVED** [review]

**Review by giansalex — APPROVED** [review]

**Pull request merged**
